x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java

```diff
@@ -319,45 +320,78 @@ private void runJob(JobTask jobTask, JobState jobState, OpenJobAction.JobParams
             }
         });
 
-        hasRunningDatafeedTask(jobTask.getJobId(), hasRunningDatafeedTaskListener);
+        getRunningDatafeed(jobTask.getJobId(), getRunningDatafeedListener);
     }
 
     private void failTask(JobTask jobTask, String reason) {
+        auditor.error(jobTask.getJobId(), reason);
         JobTaskState failedState = new JobTaskState(JobState.FAILED, jobTask.getAllocationId(), reason);
-        jobTask.updatePersistentTaskState(
-            failedState,
-            ActionListener.wrap(
-                r -> logger.debug(() -> new ParameterizedMessage("[{}] updated task state to failed", jobTask.getJobId())),
-                e -> {
-                    logger.error(
-                        new ParameterizedMessage(
-                            "[{}] error while setting task state to failed; marking task as failed",
-                            jobTask.getJobId()
-                        ),
-                        e
-                    );
-                    jobTask.markAsFailed(e);
-                }
-            )
-        );
+        jobTask.updatePersistentTaskState(failedState, ActionListener.wrap(r -> {
+            logger.debug(() -> new ParameterizedMessage("[{}] updated task state to failed", jobTask.getJobId()));
+            stopAssociatedDatafeedForFailedJob(jobTask.getJobId());
+        }, e -> {
+            logger.error(
+                new ParameterizedMessage("[{}] error while setting task state to failed; marking task as failed", jobTask.getJobId()),
+                e
+            );
+            jobTask.markAsFailed(e);
+        }));
```

Comment:
We are blowing away the task here with `jobTask.markAsFailed(e);`

So, does it make sense to also `stopAssociatedDatafeedForFailedJob(jobTask.getJobId());` when we failed to set the job to failed?

Otherwise, we have a datafeed task around but the job task is gone.

Reply:
Yes, sure, I'll add `stopAssociatedDatafeedForFailedJob` here too.

I am not sure if it will make any difference in practice though, because the only likely reason for failing to update the persistent task status is if the node is shutting down, and if that's happening then stopping the associated datafeed will also fail.